src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,201 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "compression/async_stream_zstd.h"
+
+#include "bytes/bytes.h"
+#include "bytes/details/io_allocation_size.h"
+#include "compression/logger.h"
+#include "likely.h"
+#include "units.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+#include <fmt/format.h>
+
+#include <array>
+#include <zstd.h>
+#include <zstd_errors.h>
+
+namespace compression {
+[[gnu::cold]] static void throw_zstd_err(size_t rc) {
+    if (rc == ZSTD_error_memory_allocation) {
+        ss::throw_with_backtrace<std::bad_alloc>();
+    } else {
+        ss::throw_with_backtrace<std::runtime_error>(
+          fmt::format("ZSTD error:{}", ZSTD_getErrorName(rc)));
+    }
+}
+
+static void throw_if_error(size_t rc) {
+    if (unlikely(ZSTD_isError(rc))) {
+        throw_zstd_err(rc);
+    }
+}
+
+struct zstd_workspace {
+    mutex compress_mtx;
+    mutex decompress_mtx;
+
+    ZSTD_CCtx* compress_ctx;
+    ZSTD_DCtx* decompress_ctx;
+
+    std::unique_ptr<char[], ss::free_deleter> d_workspace;
+    std::unique_ptr<char[], ss::free_deleter> c_workspace;
+
+    ss::temporary_buffer<char> d_buffer;
+    ss::temporary_buffer<char> c_buffer;
+};
+
+static thread_local zstd_workspace workspace;
+static thread_local bool workspace_initalized = false;
+
+static constexpr size_t default_decompression_size = 2_MiB;
+static constexpr int default_compression_level = 3;
+
+void async_stream_zstd::init_workspace(size_t decompression_size) {
+    if (workspace_initalized) {
```

Comment:
This pattern is a little worrisome. We only init on the first call but the init behavior depends on `decompression_size` right?

So if it is called a second time with a larger value, we just silently drop that request. A simple "safe" solution would be to assert the the value is <= the value actually used, so we die otherwise.

Reply:
Good point, added assert to ensure that the requested allocation size is never greater than the initial allocation.

Reply:
Ideally we avoid this entirely by changing e.g., how the initialization happens, e..g, you pass this value while constructing some "context" object which then lets you do decompressions with the passed values, but maybe that's not possible with this global singleton pattern.